### ISSUE TITLE
[cluster-autoscaler] CriticalAddonsOnly taint ignored

### ISSUE THREAD

**Issue by jclangst** [kubernetes/autoscaler#4097]
**Which component are you using?**:

cluster-autoscaler

**What version of the component are you using?**:

1.20.0

Component version:

**What k8s version are you using (`kubectl version`)?**:

1.20.5

<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.6", GitCommit:"8a62859e515889f07e3e3be6a1080413f17cf2c3", GitTreeState:"clean", BuildDate:"2021-04-15T03:28:42Z", GoVersion:"go1.15.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.5", GitCommit:"54684493f8139456e5d2f963b23cb5003c4d8055", GitTreeState:"clean", BuildDate:"2021-03-22T23:02:59Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}

</pre></details>

**What environment is this in?**:

Azure w/ VMSS scaling configured via auto-discovery

**What did you expect to happen?**:

When I apply taint `CriticalAddonsOnly=true:NoSchedule` to a node group, it should be respected.

**What happened instead?**:

This particular taint is not respected during scaling calculations.

**How to reproduce it (as minimally and precisely as possible)**:

Create a node group with this taint and one without, scale a deployment without a matching toleration, notice that the tainted node group will get scaled occasionally.

**Anything else we need to know?**:

This appears to be a result of this taint filtering logic; https://github.com/kubernetes/autoscaler/blob/d33cc1bc400f6d5cccaa0ca95696fa0e1780df29/cluster-autoscaler/utils/taints/taints.go#L70

I believe this is the problem in this old issue that went unresolved: https://github.com/kubernetes/autoscaler/issues/2434

After searching the repository, I could not find any references to this value in the core logic. While I may be missing something, I am not sure what the intended purpose of this filter is.  The only thing I can find is a reference to a "rescheduler". I think it may be related to this KEP but I'm not sure if its still relevant: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/rescheduling.md

It is hindering my use case in the Azure/AKS b/c that is the _only_ taint that can be added to the default system node pool. Frequently the scaling is _very_ delayed as the autoscaler will scale this node pool incorrectly and repeatedly. My hacky workaround is to just have the autoscaler ignore this node pool, but I am afraid this unexpected behavior will continue to surprise people. Moreover, I'd like to be able to not ignore this pool in the future.

(1) Can it be removed?
(2) If not, can a warning and explanation be added to the FAQ?

**Comment by stevehipwell** [comment]
We've just discovered this issue, are there any plans to resolve or document it?

**Comment by jclangst** [comment]
I think you may have misread. The documentation you linked references a Kubernetes component called the "rescheduler" that does not exist, at least in modern kubernetes distributions. To my knowledge the modern Kubernetes scheduler never ignores node taints so it does not make sense for the autoscaler to ignore this one taint in particular. It looks to be legacy code built in a time when there were proposals to make something called the "rescheduler" a main Kubernetes component. Since that did not happen, this should be removed.

**Comment by pierluigilenoci** [comment]
The only official documentation on Critical Add-On that I could find is this: 
https://kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/

**Comment by jclangst** [comment]
Yep! There is that, but the scheduler still is not going to ignore node taints even for `systen-X-critical` priority classes. Ignoring this node taint still does make sense even in that context.

**Comment by jclangst** [comment]
Sorry. Closed wrong issue!

**Comment by pierluigilenoci** [comment]
Small update. AWS is also in the same boat. 
Ref: https://github.com/aws/containers-roadmap/issues/1389

**Comment by pierluigilenoci** [comment]
@jclangst @stevehipwell any update?

**Comment by stevehipwell** [comment]
@gjtempleton would you take a PR to remove this undocumented taint from the CA logic?

**Comment by robinkb** [comment]
I am blocked by the same issue, and have to fork cluster-autoscaler to remove this special case to get my clusters to work the way I expect them to.

I would _really_ appreciate at least an option to toggle this behavior.

**Comment by pierluigilenoci** [comment]
@gjtempleton any news about this?

**Comment by pierluigilenoci** [comment]
@ib-ak I'm sorry, but I can't help you much. 
Like you, I, too, am waiting for this problem to be solved.

**Comment by sftim** [comment]
BTW, `CriticalAddonsOnly` is *not* an official Kubernetes taint; if it were, you'd see it listed in [Well-Known Labels, Annotations and Taints](Well-Known Labels, Annotations and Taints) and it would be prefixed, eg with `kubernetes.io/`.

For example, nodes can come online tainted as `node.cloudprovider.kubernetes.io/uninitialized`. That's a real, registered taint and the prefix makes it public. Other organizations can register their own public taints with a prefix they control.

Using a **private** taint, `CriticalAddonsOnly`, is unhelpful and cluster operators should try not to do so. On the other hand, cluster-autoscaler is generally available.

I think that cluster-autoscaler's use of this private taint represents an architecture bug.

**Comment by vadasambar** [comment]
Came to know about this issue today because of https://kubernetes.slack.com/archives/C09R1LV8S/p1685972225498489

I am not sure if I have fully grasped the details here. I have added it as an item on the agenda for the upcoming sig meeting to decide what to do about this. Once we reach a decision, I would be happy to raise a PR (I am working on 2 other PRs so some delay should be expected).

**Comment by vadasambar** [comment]
Link to agenda for upcoming sig meeting: https://docs.google.com/document/d/1RvhQAEIrVLHbyNnuaT99-6u9ZUMp7BfkPupT2LAZK7w/edit#bookmark=id.e9v2tiuubvsq

**Comment by discordianfish** [comment]
My 2 cents: Since it's not an official taint, it shouldn't be treated different than any other taint. Therefor I'd remove it.
That being said, if @stevehipwell is right:
> AWS EKS uses this taint when starting new nodes, which I suspect is to allow the aws-node CNI pod to start before allowing other workloads to schedule, and then removes it after the node is ready (has a CNI configured). 

...this will cause problems as described above. But frankly this shouldn't be too much of a concern for this project. Might require a breaking/major release but EKS is in the wrong using this taint instead of the official e.g `node.kubernetes.io/network-unavailable`

**Comment by sftim** [comment]
If preferred, EKS could also mint a taint such as `eks.aws/critical-addons-unavailable`. Same for other providers, which helps if the meanings and implications aren't 100% congruent.

**Comment by pierluigilenoci** [comment]
@discordianfish If you hope cloud providers will use "official" methods, you are naïve and wait forever. 😜 
But this thing has to be solved somehow. 🤷🏻‍♂️ 
Maybe it could be planned to configure a list of custom taints. 💡 
So that the default tool uses the correct taints, but you can bend the operation to use custom ones to follow the "creative flair" of the cloud provider developers. 🤔

**Comment by discordianfish** [comment]
@pierluigilenoci Precisely my point, this project should assume the 'official' methods and leave it to downstream to converge to that (aka break it)

**Comment by stevehipwell** [comment]
@discordianfish I'm not 100% sure if EKS still uses this taint but I would say they're more correct than AKS who use it as their system node taint. AFAIR the `CriticalAddonsOnly` was used in an official or semi-official way for the same purpose as EKS is/was using it.

**Comment by pierluigilenoci** [comment]
@discordianfish The project must assume the "official" methods as default behavior but could give the possibility to use "custom" methods on request (for example, through some configuration).

@stevehipwell between AKS and EKS, they have an internal competition over who does things less orthodox. 🤣

**Comment by vadasambar** [comment]
Brought this up in the sig meeting yesterday (12th June 2023). Let's wait and see if we get a response from someone in the community who has historical context on this. Otherwise I will bring this up again in the upcoming sig meeting (Mon, 19th June 2023)

**Comment by jclangst** [comment]
Can you be more specific about what historical context you are looking for? We have already done a deep dive in the current and historical codebases [here](https://github.com/kubernetes/autoscaler/issues/4097#issuecomment-906525705).

**Comment by vadasambar** [comment]
@jclangst I basically want someone from CA community with better historical context (than me) around why we haven't removed this taint yet and if it's okay to just go ahead and remove it. Based on the community's response, we either remove the taint completely from the code OR have an option to preserve the current behavior while giving an option to change the behavior based on the use-case (maybe using a flag)

**Comment by jclangst** [comment]
@vadasambar To that end, I would also be interested in what is holding us up from choosing a path on this.

**Comment by jclangst** [comment]
Given that the CA already has flags for ignoring certain taints, it seems to me implicit + immutable ignore behavior should be removed. It might generate some very minor work for cloud providers that have relied on the implicit, undocumented behavior, but as long as this is coupled with a major version bump that seems completely reasonable (in my personal opinion).

**Comment by MaciekPytel** [comment]
Contrary to what comments above say - it's not a private extension, CriticalAddonsOnly _was_ an official Kubernetes taint (to the extent anything was official in those days - it was used by core k8s code which I think was how we defined "official" back in 2017 :) ). 
It's not in well-known labels and it doesn't have kubernetes.io/ prefix, because it predates organizing 
well-known labels or standardizing on prefix (same reason why CA's own to be deleted taint doesn't use prefix).

This taint was used by 'rescheduler' which very much existed and wasn't just a proposal. I don't remember the details after all those years, but back than pod priority didn't exist and daemonset pods were scheduled directly by DaemonsetController, not by scheduler. IIRC Rescheduler was basically a separate controller doing equivalent of today pod preemption for system pods. Since there was no pod priority back than it had to temporarily taint the node while performing the preemption, otherwise any pod could schedule using the capacity freed by preemption. The only way to guarantee the intended pod would use the capacity was to temporarily taint a node so that non-system pods couldn't schedule there. Once the preemption was done, the taint would be removed. I'm sure I'm getting some details wrong, but I think that was the general idea.
Rescheduler seems to have been removed in 1.11: https://github.com/kubernetes/kubernetes/pull/67687.

So yeah - this is very much legacy. I don't think CA should be required to maintain support for a feature removed from kubernetes back in 2018 and I'd be happy to approve a PR removing it.

Now - the issue brought on sig autoscaling seems to have been triggered by the idea of using this particular taint to mark control plane nodes. I don't think that's consistent with how CriticalAddonsOnly taint was originally intended to be used and I guess that could be considered a private extension (using an "official" taint for a different purpose). I have no idea what platforms out there may be using and whether they will be impacted by the change and it will technically be backward incompatible change, so I'd ask anyone making this change to make sure to include a release note that clearly calls out the backward-incompatible change in behavior.

**Comment by sftim** [comment]
I'm guessing `CriticalAddonsOnly` became a private taint when we introduced the registry for well known ones (Apr 20, 2017 according to Git history). We didn't go around and look for legacy uses across the project at that point.

As it's been around so long, I strongly agree on the value of some kind of deprecation process.

**Comment by vadasambar** [comment]
https://github.com/kubernetes/autoscaler/pull/5838 is ready for review.
